File: client-apps/cli/internal/cli/daemon/reset.go
package daemon

import (
	"os"
	"path/filepath"

	"github.com/pkg/errors"
	"github.com/rs/zerolog/log"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/config"
	"github.com/stigmer/stigmer/client-apps/cli/internal/cli/temporal"
)

// ResetOptions controls the scope of a reset operation.
type ResetOptions struct {
	IncludeConfig bool // Also remove config.yaml (user must re-run setup wizard)
}

// ResetResult describes what was removed during a reset.
type ResetResult struct {
	ServicesStopped bool
	RemovedPaths    []string
}

// Reset stops all services and removes runtime state from configDir (~/.stigmer).
//
// By default, config.yaml is preserved so the user does not lose API keys or
// backend preferences. Pass IncludeConfig to remove it as well.
func Reset(configDir, dataDir string, opts ResetOptions) (*ResetResult, error) {
	result := &ResetResult{}

	if err := stopAllServices(dataDir); err != nil {
		log.Debug().Err(err).Msg("Stop returned an error (may already be stopped)")
	}
	result.ServicesStopped = true

	removers := []struct {
		name string
		fn   func(string) ([]string, error)
	}{
		{"data directory", removeDataDir},
		{"temporal state", removeTemporalState},
		{"downloaded binaries", removeDownloadedBinaries},
		{"root logs", removeRootLogs},
	}

	for _, r := range removers {
		paths, err := r.fn(configDir)
		if err != nil {
			return result, errors.Wrapf(err, "failed to remove %s", r.name)
		}
		result.RemovedPaths = append(result.RemovedPaths, paths...)
	}

	if opts.IncludeConfig {
		paths, err := removeConfigFile(configDir)
		if err != nil {
			return result, errors.Wrap(err, "failed to remove configuration")
		}
		result.RemovedPaths = append(result.RemovedPaths, paths...)
	}

	return result, nil
}

// stopAllServices attempts to stop all running services via daemon.Stop.
// Errors are intentionally non-fatal: the server may already be stopped.
func stopAllServices(dataDir string) error {
	if !IsRunning(dataDir) {
		return nil
	}
	return Stop(dataDir)
}

// removeDataDir removes the entire data directory (~/.stigmer/data).
func removeDataDir(configDir string) ([]string, error) {
	dataDir := filepath.Join(configDir, config.DefaultDataDir)
	return removeIfExists(dataDir)
}

// removeTemporalState removes Temporal data directory and state files.
func removeTemporalState(configDir string) ([]string, error) {
	targets := []string{
		filepath.Join(configDir, "temporal-data"),
		filepath.Join(configDir, temporal.TemporalPIDFileName),
		filepath.Join(configDir, temporal.TemporalLockFileName),
	}
	return removeAll(targets)
}

// removeDownloadedBinaries removes the bin directory with downloaded tools.
func removeDownloadedBinaries(configDir string) ([]string, error) {
	binDir := filepath.Join(configDir, "bin")
	return removeIfExists(binDir)
}

// removeRootLogs removes root-level log directory and LLM PID file.
func removeRootLogs(configDir string) ([]string, error) {
	targets := []string{
		filepath.Join(configDir, "logs"),
		filepath.Join(configDir, "llm.pid"),
	}
	return removeAll(targets)
}

// removeConfigFile removes config.yaml.
func removeConfigFile(configDir string) ([]string, error) {
	configFile := filepath.Join(configDir, config.ConfigFileName)
	return removeIfExists(configFile)
}

// removeIfExists removes a path and returns it if it existed.
func removeIfExists(path string) ([]string, error) {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return nil, nil
	}
	if err := os.RemoveAll(path); err != nil {
		return nil, errors.Wrapf(err, "failed to remove %s", path)
	}
	return []string{path}, nil
}

// removeAll removes multiple paths, collecting those that existed.
func removeAll(paths []string) ([]string, error) {
	var removed []string
	for _, p := range paths {
		r, err := removeIfExists(p)
		if err != nil {
			return removed, err
		}
		removed = append(removed, r...)
	}
	return removed, nil
}

File: client-apps/cli-ts/src/local/reset.ts
// `stigmer reset` — stop the stack and wipe all persistent local state, so the
// next `up` starts from a clean slate. Configuration (config.yaml) is preserved
// by default so users do not lose API keys; `includeConfig` removes it too.
//
// The set of removed paths is a faithful port of the Go CLI's reset, spanning
// both roots of the `~/.stigmer` layout (the SQLite db + Temporal state under
// the config dir, and everything under the data dir).

import { existsSync, rmSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { configPath } from "../config/paths.js";
import {
  binDir,
  configDir,
  databasePath,
  dataDir,
  storagePath,
  temporalDataDir,
  temporalLockFile,
  temporalPidFile,
} from "./paths.js";
import { down, isRunning } from "./daemon/launch.js";

export interface ResetOptions {
  /** Also remove config.yaml (the user must reconfigure on next start). */
  includeConfig?: boolean;
}

export interface ResetResult {
  servicesStopped: boolean;
  removedPaths: string[];
}

/** Stop the stack if running. Returns whether anything was stopped. Injectable
 * so reset's file-wiping can be tested without touching a real daemon. */
export type StopFn = (home: string) => Promise<boolean>;

const defaultStop: StopFn = async (home) => {
  if (!(await isRunning(home))) return false;
  await down(home);
  return true;
};

/** Stop services (if running) and remove persistent state under `home`. */
export async function reset(
  options: ResetOptions = {},
  home: string = homedir(),
  stop: StopFn = defaultStop,
): Promise<ResetResult> {
  const servicesStopped = await stop(home);

  const config = configDir(home);
  const db = databasePath(home);

  // Order: db (+ sidecars) first, then the data tree, then config-dir state.
  const targets = [
    db,
    `${db}-wal`,
    `${db}-shm`,
    dataDir(home),
    storagePath(home),
    join(config, "sessions"),
    join(config, "runtimes"),
    temporalDataDir(home),
    temporalPidFile(home),
    temporalLockFile(home),
    binDir(home),
    join(config, "logs"),
    join(config, "llm.pid"),
  ];

  const removedPaths: string[] = [];
  for (const target of targets) {
    if (existsSync(target)) {
      rmSync(target, { recursive: true, force: true });
      removedPaths.push(target);
    }
  }

  if (options.includeConfig === true) {
    const cfg = configPath(home);
    if (existsSync(cfg)) {
      rmSync(cfg, { force: true });
      removedPaths.push(cfg);
    }
  }

  return { servicesStopped, removedPaths };
}
